src/Kestrel.Core/Internal/Infrastructure/SafePipeWriterFlusher.cs

```diff
@@ -0,0 +1,76 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO.Pipelines;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Connections;
+using Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure
+{
+    public class SafePipeWriterFlusher
```

Comment:
Safe from what? ThreadSafe?

Reply:
It is ThreadSafe, but that's only part of it. Unlike calling PipeWriter.FlushAync directly, you can have multiple callers await the same flush task. The following written directly agains a PipeWriter would throw an invalid operation exception whenever there was actually backpressure:

```c#
await Task.WhenAll(_writer.FlushAsync().AsTask(), _writer.FlushAsync.AsTask())
```

The same code written against SafePipeWriterFlusher would work. I can't really think of a great reason anyone would want to have multiple awaiters Response.Body.FlushAsync(), but it's something we support for HTTP/1.1 so I copied the code to make that work and reused it.

I'll call it StreamSafePipeFlusher to indicate that it's safe to use to implement Stream.FlushAsync.